paasd/src/service.rs

```diff
@@ -27,25 +60,64 @@ impl server_types::ProcessService for ProcessService {
     type GetLogsStream =
         Pin<Box<dyn Stream<Item = Result<LogsResponse, tonic::Status>> + Send + Sync + 'static>>;
 
-    async fn exec(&self, _req: Request<ExecRequest>) -> Result<Response<ExecResponse>, Status> {
-        Err(Status::unimplemented(""))
+    async fn exec(&self, req: Request<ExecRequest>) -> Result<Response<ExecResponse>, Status> {
+        let uid = Self::authenticate(&req)?;
+        let req = req.into_inner();
+        if req.args.is_empty() {
+            return Err(Status::invalid_argument("'args' must be a non-empty array"));
+        }
+        match Process::spawn(&*req.args[0], req.args[1..].iter().map(AsRef::as_ref)) {
+            Ok(p) => {
+                let pid = self.store.insert(&uid, p);
+                Ok(Response::new(ExecResponse {
+                    id: Some(pid.into()),
+                }))
+            }
+            Err(e) => Err(Status::unknown(format!("{}", e))),
+        }
     }
 
     async fn get_logs(
         &self,
-        _req: Request<LogsRequest>,
+        req: Request<LogsRequest>,
     ) -> Result<Response<Self::GetLogsStream>, Status> {
-        Err(Status::unimplemented(""))
+        let uid = Self::authenticate(&req)?;
+        let req = req.into_inner();
+        let pid = req
+            .id
+            .ok_or_else(|| Status::invalid_argument("Process ID not given"))?;
+        let process = self.get_process(pid, &uid)?;
+        // TODO: buffer multiple lines
```

Comment:
`LogsResponse` can return multiple lines (`Vec<Bytes>`). This would likely reduce overhead, at least on the initial fetch of logs (where a large backlog could have already been accumulated).